# From: Transformers and Text-Generation
by Liam Dugan (UPenn). 


Please write your answers and code in the cells with questions below. 

----------

For this homework, we will take ideas from the entire class: language models, text generation, vector-based word representations, syntactic analysis, and neural networks. We'll be using large, pre-trained language models to generate text, and studying how we can fine-tune these large language models to generate text in whatever genre and style we want!

In this assignment you will get:
1. An overview of the "Transformer" architecture is and why it is particularly well suited for Natural Language Processing tasks
2. An introduction to the Generative Pretrained Transformer (GPT) family, which is a set of large-scale language models that can be used to generate text that often sounds like it was written by a human.
3. Experience with using the HuggingFace package to fine-tune these models to generate text that sounds like it comes from a specific source.

# Part 1: What is a Transformer? (Reading)
<figure align="center">
<img src="https://media.giphy.com/media/VeWllmR9zfaco/giphy.gif" />
<figcaption>(It's probably not this guy, right?)</figcaption>
</figure>

### The Transformer

The current state-of-the-art for a variety of natural language processing tasks belongs to the **Transformer** architecture, first published December 6th 2017. 

The Transformer can be thought of as a big feed-forward network with every feed-forward layer containing something called an "attention module". 

>You might be wondering: why are we moving back to feed-forward networks after having so much success with recurrent neural networks and variants like LSTMs? Aren't RNNs naturally poised to handle sequences as their inputs? Well, as it turns out, the sequential nature of RNNs make them really difficult to train in a distributed/parallel fashion. So while RNNs make more sense to use on sequences of inputs, serial networks such as the transformer can be trained much faster, allowing orders of magnitude more training data to be used. 



### Reading \# 1 - What is a Transformer?

In order to get a good grasp on exactly *why* these models are so good it's important to understand what they are and how they work. 

Your first task for this homework is to read the blog post ["The Illustrated Transformer" by Jay Alammar](http://jalammar.github.io/illustrated-transformer/). This blog post explains the transformer architecture (and the all-important "Attention Module") with helpful visualizations and diagrams. 

**You should read this post very closely and understand exactly what the Transformer is and how it works. Once you're finished reading, answer the following questions in 2-3 sentences each.**

1. (2 pts) What is Self-Attention (at a high level)?

   > As the model processes each word (each position in the input sequence), self-attention allows it to look at other positions in the input sequence for clues that can help lead to a better encoding for this word. It is the method the Transformer uses to bake the “understanding” of other relevant words into the one we’re currently processing.

2. (2 pts) How is Self-Attention computed?

   > $$Atten = \frac {softmax(Q*K.T)} {\sqrt{d_k}} * V$$

3. (2 pts) What do the "Query", "Key", and "Value" vectors encode (at a high level)?

   > When we compute Self-Attention, we need to score each word of the input sentence against this word. The score is calculated by taking the dot product of the "Query" vector with the "Key" vector of the respective word we’re scoring. Then, we will multiply each "Value" vector by the softmax score. It means keep intact the values of the words we want to focus on, and drown-out irrelevant words.

4. (2 pts) What is an attention "head" and why should we use multiple heads?

   > An attention "head" means we do self-attention once, muliple heads means we do self-attention multiple times. Using multiple heads expands the model’s ability to focus on different positions and gives the attention layer multiple “representation subspaces”.

5. (2 pts) What are positional embeddings?

   > Positional embeddings represent the order of the words in the input sequence. These vectors follow a specific pattern that the model learns, which helps it determine the position of each word, or the distance between different words in the sequence

6. (2 pts) Why are positional embeddings important?

   > Because the self-attention do not have information about the order of words, but the order of words is important, such as "I love you" and "you love me" are total different.




-----------
### Reading \# 2 - Transformer Language Models

On June 11th 2018, OpenAI released a model named **GPT**, standing for *Generative Pre-Trained Transformer*. This model was a Transformer architecture that was modified such that it could be used for Text Generation instead of sequence to sequence modeling. This model was also pre-trained, which means that anyone could download the fully trained model and use it without needing to train the model themselves. 

On February 14th 2019, OpenAI released a blog post detailing a brand new version of GPT that had an insane **1.5 billion parameters**. They named this version **GPT-2**. To train such a large model, OpenAI crawled 40GB worth of text from the web (roughly 20,000,000,000 words). 

GPT-2 is an extremely impressive language model that can generate text that often sounds so plausible that it seems like it might have been written by a human.  Here is an example of what GPT-2 can generate, taken from [OpenAI’s blog post](https://openai.com/blog/better-language-models/):

**Human-Written Prompt:**
> “Miley Cyrus was caught shoplifting from Abercrombie and Fitch on Hollywood Boulevard today.”

**Model Continuation:**
> The 19-year-old singer was caught on camera being escorted out of the store by security guards.
> 
> The singer was wearing a black hoodie with the label ‘Blurred Lines’ on the front and ‘Fashion Police’ on the back.
> 
> Scroll down for video
> 
> Shoplifting: Miley Cyrus was caught shoplifting from Abercrombie and Fitch on Hollywood Boulevard today (pictured)
> 
> The singer was also wearing a pair of black-rimmed glasses, a black jacket, black jeans and black sandals.
> 
> She was carrying a pair of black and white striped gloves and a small black bag.

*(To be clear: Miley Cyrus was never actually arrested for shoplifting)*.  

The quality of the model's output was so good, that the researchers at OpenAI were worried that, in addition to positive applications, it might be used for malicious purposes like generating fake news, impersonating people online, creating abusive or fake content to post on social media, or automating spam and phishing scams. In a [demo to Wired Magazine](https://www.wired.com/story/ai-text-generator-too-dangerous-to-make-public/), the reporter typed in “Hillary Clinton and George Soros” and GPT-2 generated a crazy rant with lots of conspiracy theories: 
> Hillary Clinton and George Soros are a perfect match, and their agenda appears to be to create a political movement where Soros and his political machine and Clinton are two of the only major players. This is the first time Soros and Clinton have been caught on tape directly colluding in promoting the same false narrative. One of the key revelations in the leaked audio was Clinton's admission to a Russian banker that she knew about the Uranium One deal before it was approved by Congress. Clinton was shown sharing the same talking points that were originally drafted by a Fusion GPS contractor hired by an anti-Trump Republican donor. The leaked audio is the clearest evidence yet that the Clinton campaign and the Hillary Foundation colluded with Fusion GPS to manufacture propaganda against President Trump.


They were concerned enough that they labeled GPT-2 "too dangerous to release", and OpenAI initially refused to release their dataset, training code, or GPT-2 model weights.  OpenAI decided to release in a delayed, phased fashion so that researchers could spend time working on automatic detection of generated text.

In this homework, you'll get to be the judge of how good GPT-2 is, as you'll be using it yourself to generate text!

**To start your journey into the world of Text Generation, you should read Part 1 of the blog post ["The Illustrated GPT-2" by Jay Alammar](http://jalammar.github.io/illustrated-gpt2/) and answer the following questions in 2-3 sentences each**

1. (4 pts) How does the architecture of GPT-2 differ from the standard Encoder-Decoder Transformer model?
   > GPT-2 consists of solely stacked decoder blocks from the transformer architecture. Instead of using encoder to model sequences and use decoder to generate words, GPT-2 only use decoder to model and generating. In GPT-2 masked self-attention is used : the decoder is only allowed to glean information from the prior words in the sentence (plus the word itself).
2. (4 pts) What is the difference between "Masked Self-Attention" and "Self-Attention"
   > The masked self-attention layer is only allowed to attend to earlier positions in the output sequence. This is done by masking future positions (setting them to -inf) before the softmax step in the self-attention calculation.
3. (4 pts) What are logits? How are they computed? and How does GPT-2 use them to decide which word to predict next?
   > Logits are probability distribution over all possible integer token IDs. Logits are the output of GPT-2, and then use softmax, $\frac {exp(X_i)} {sum(exp(X_i)}$, to compute probability. And predict next word with high probability.

### Aside: GPT-3 

On June 11th 2020, OpenAI released GPT-3 [(paper)](https://arxiv.org/pdf/2005.14165.pdf) [(wikipedia)](https://en.wikipedia.org/wiki/GPT-3). This model has an unfathomable **175 billion parameters** (100x larger than GPT-2!) and was trained on 570GB of text! This model is virtually indistinguishable from human output and can generate text about any topic and in any style with only a few words of priming text. It is by far the largest language model ever trained and it can do some very terrifying things.

GPT-3 Can:
- Generate JSX code off natural language descriptions
- Generate Emojis based off of descriptions of the feeling
- Generate regular expressions off natural language descriptions
- Generate website mockups off natural language descriptions
- Generate charts with titles, labels and legends from natural language descriptions
- Explain python code in plain english
- Automatically generate quiz questions (and grade them)
- Generate Latex from natural language descriptions
- Generate Linux commands from natural language descriptions
- Generate a Machine Learning model from natural language descriptions

[Here's a collection of 21 things GPT-3 can do (with examples)](https://machinelearningknowledge.ai/openai-gpt-3-demos-to-convince-you-that-ai-threat-is-real-or-is-it/#OpenAI_GPT-3_Demos)

[Here's a NYT article about how GPT-3 can write code, poetry, and argue](https://www.nytimes.com/2020/11/24/science/artificial-intelligence-ai-gpt3.html)

[Here's an article GPT-3 wrote for The Guardian about how it loves humans and would never subjugate humanity](https://www.theguardian.com/commentisfree/2020/sep/08/robot-wrote-this-article-gpt-3)

**You may optionally choose to read Jay Alammar's most recent blog post ["How GPT3 Works - Visualizations and Animations"](http://jalammar.github.io/how-gpt3-works-visualizations-animations/) from July 2020 if you're curious as to how GPT-3 differs from GPT-2**

Similarly to GPT-2, OpenAI has decided not to release GPT-3, this time opting to put GPT-3 behind an API which you need to request permission to use. This allows them to control exactly who can generate text and what type of text is generated. While this is a good solution in the short term, the long term implications of GPT-3 are still unclear.

If you are interested in trying out GPT-3 yourself, feel free to [Join the OpenAI API Waitlist](https://share.hsforms.com/1Lfc7WtPLRk2ppXhPjcYY-A4sk30)

-------------------------------

# Part 2: GPT-2 Text Generation with HuggingFace

Phew, that was a lot of reading. Now lets get to the fun part! Let's use the transformer to generate some text!!

We will use the [Transformers library from HuggingFace](https://transformer.huggingface.co), which provides support for many Transformer-based language models like GPT-2. 

**IMPORTANT: Make sure that you have GPU set as your Hardware Accelerator in `Runtime > Change runtime type` before running this Colab.**

In [1]:
!pip install transformers

In [2]:
!nvidia-smi

Sun Apr 11 04:09:00 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.67       Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   35C    P0    24W / 300W |      0MiB / 16160MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [3]:
import logging
import transformers
logging.getLogger("transformers.generation_utils").setLevel(logging.ERROR)  # No warning on sample size
logging.basicConfig(
        format="%(message)s",
        datefmt="%m/%d/%Y %H:%M:%S",
        level=logging.INFO,
)
logger = logging.getLogger(__name__)

## 2.1 The 'Pipeline' Interface

The simplest way to use the HuggingFace library is to use their [Pipeline interface](https://huggingface.co/transformers/main_classes/pipelines.html)

There are many different types of Pipelines available but in this section we'll use the TextGenerationPipeline to get up and running with pretrained gpt2 as fast as possible

In [4]:
from transformers import pipeline

In [5]:
# Note: device=0 means to use GPU, device=-1 is to use CPU
generator = pipeline('text-generation', model='gpt2', device=0) 

In [6]:
outputs = generator('I love cats.')
print(outputs[0]['generated_text'])

I love cats. I am extremely allergic to them. I am a vet, and since I love cats, I've spent the last few days just making up for all the other things I have done for cats.

But now, instead of


Note that the 'text-generation' pipeline will work with any **auto-regressive** language model (a.k.a 'causal-lm' models according to the HuggingFace lingo). You can find a list of all such models here https://huggingface.co/models?filter=causal-lm. 

(6 pts) **Your first task is to use the Pipeline interface to get generation output below for at least two different 'causal-lm' models (One of these two can be a different version of GPT2, but make sure at least one is a non-gpt family language model)**

In [7]:
## YOUR CODE HERE FOR MODEL 1
generator_1 = pipeline('text-generation', model='xlnet-base-cased', device=0) 

In [8]:
outputs = generator_1('I love cats.')
print(outputs[0]['generated_text'])

I love cats. I’m not one for people giving them away, but I’m the kind who’ll give them to someone who wants to adopt them. I want to live a life of sustaining a family (without their children) for a long period of time after they grow old and their children are allowed to leave their parents), but I really do want to be a good mother to someone who


In [9]:
## YOUR CODE HERE FOR MODEL 2
generator_2 = pipeline('text-generation', model='microsoft/DialoGPT-medium', device=0) 

In [10]:
outputs = generator_2('I love cats.')
print(outputs[0]['generated_text'])

I love cats.They are so cute .


In [11]:
# nlp = pipeline("question-answering")

# context = "Extractive Question Answering is the task of extracting an answer from a text given a question. An example of a question answering dataset is the SQuAD dataset, which is entirely based on that task. If you would like to fine-tune a model on a SQuAD task, you may leverage the `run_squad.py`."

# print(nlp(question="What is extractive question answering?", context=context))
# print(nlp(question="What is a good example of a question answering dataset?", context=context))

## 2.2 Dissecting the Pipeline
Now that was easy!

As beautiful and easy as the Pipeline interface is, we want to know what's going on under the hood!

There are four main steps to a text generation pipeline:
1. (Tokenize) Turn the raw input text into a vector of integer token IDs using a tokenizer

2. (Encode) Feed those token IDs into the language model by querying for each token's embedding in the model's embedding matrix (the "encoder") and then feed the "encoded" sequence into the decoder module

3. (Decode) The decoder will output logits (a probability distribution over all possible integer token IDs) and we sample from those logits to get our next token -- repeat until EOS token is generated or we hit max_length

4. (Detokenize) Take the output sequence of token IDs and turn them from integer token IDs back to tokens with the tokenizer

Below you'll see how HuggingFace does this:

First we have to initialize both the tokenizer and the model from their pre-trained checkpoints. Note that the tokenizer has to match the model.

In [12]:
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained('gpt2')
model = AutoModelForCausalLM.from_pretrained('gpt2').cuda()

In [13]:
#### Step 1: Tokenize the input into integer token IDs
inputs = tokenizer.encode("Hello, how are you?", return_tensors='pt').to(model.device)
print("Input Token IDs: " + str(inputs))

Input Token IDs: tensor([[15496,    11,   703,   389,   345,    30]], device='cuda:0')


In [14]:
#### Step 2 and 3: Feed in the integer token IDs and get out a sequence of token IDs as output
outputs = model.generate(inputs)
print("Output Token IDs: " + str(outputs))

Output Token IDs: tensor([[15496,    11,   703,   389,   345,    30,   198,   198,    40,  1101,
           257,  1310,  1643,   286,   257, 34712,    13,   314,  1101,   257]],
       device='cuda:0')


In [15]:
#### Step 4: Feed in the integer token IDs and get out a sequence of token IDs as output
output_text = [tokenizer.decode(x) for x in outputs]
print("Output Text: " + str(output_text))

Output Text: ["Hello, how are you?\n\nI'm a little bit of a nerd. I'm a"]


Now that you have dissected the pipeline, it's time to play with some common parameters!

[Check out this demo notebook from HuggingFace](https://github.com/huggingface/blog/blob/master/notebooks/02_how_to_generate.ipynb) for a good overview of the different generation parameters and what they do (with example code!).

The full documentation on all of the parameters you can use in the generate function can be found [here](https://huggingface.co/transformers/main_classes/model.html#transformers.generation_utils.GenerationMixin.generate)

As an example, below we have a call to generate that:
- randomly samples from the top 50 words in the output distribution (rather than just greedily picking the best one every time)
- downweights the probability of all previously generated tokens by a factor of 1.2 (to prevent repetition)
- goes on for 512 tokens, because its more interesting

In [16]:
inputs = tokenizer.encode("The Los Angeles Lakers", return_tensors='pt').to(model.device)
outputs = model.generate(
      inputs,
      do_sample=True,          # Randomly sample from the logits instead of greedily picking next word with highest probability
      top_k=50,                 # Only sample from the top 50 most likely words
      repetition_penalty=1.2,    # Downweights the probability of all previously generated tokens by a factor of 1.2
      max_length=512          # Generate for a maximum of 512 tokens
  )
print(tokenizer.decode(outputs[0], skip_special_tokens=True).replace('\n',' '))


The Los Angeles Lakers, who have long taken pride in playing for the "Unsung Hero" of their sport (and are currently being praised by many players around town), will be looking to play a very interesting game on Saturday. - ESPN's Chris Cillizza covers college basketball and its importance with three hours per day starting at 3pm Pacific Time from San Francisco State University where he shares how his coaches approach these games every week or so along their season ticket list that comes each weekday afternoon:


**Your job is to provide two different examples of generation output from GPT-2 with different choices of generation parameters. You must also provide a 1-2 sentence explanation of what these parameters do and how they affect your output**

Feel free to get creative with this! Really poke around and try to find the combination of settings that gives you the best sounding text! The ways in which these parameters affect how 'human-like' a section of generated text sounds is an area of active research. :)

(4 pts) YOUR ANSWER HERE - EXPLANATION FOR HPARAM VARIATION 1
> Using beam search, with 10 beams, controling repeat ngram size less than 2, and use early stopping

In [17]:
inputs = tokenizer.encode("The Los Angeles Lakers", return_tensors='pt').to(model.device)
outputs = model.generate(
      inputs,
      max_length=50, 
      num_beams=10, 
      no_repeat_ngram_size=2,
      num_return_sequences=5, 
      early_stopping=True
  )
# now we have 3 output sequences
print("Output:\n" + 100 * '-')
for i, beam_output in enumerate(outputs):
    print("{}: {}\n".format(i, tokenizer.decode(beam_output, skip_special_tokens=True)).replace('\n',' '))


Output:
----------------------------------------------------------------------------------------------------
0: The Los Angeles Lakers are the only team in the NBA that has won at least 100 games in each of the past three seasons.  The Lakers have won three of their past four games at Staples Center, including a 108-98 victory over the 
1: The Los Angeles Lakers are the only team in the NBA that has won at least 100 games in each of the past three seasons.  The Lakers have won three of their past four games at Staples Center, including a 103-98 victory over the 
2: The Los Angeles Lakers are the only team in the NBA that has won at least 100 games in each of the past three seasons.  The Lakers have won three of their past four games at Staples Center, including a 103-99 victory over the 
3: The Los Angeles Lakers are the only team in the NBA that has won at least 100 games in each of the past three seasons.  The Lakers have won three of their past four games at Staples Center, includin

(4 pts) YOUR ANSWER HERE -- EXPLANATION FOR HPARAM VARIATION 2
> Use top_p sample with top_k sample, and use temperature=1, reutrn 3 sequences.

In [18]:
## YOUR CODE HERE FOR HYPERPARAMETER VARIATION 2
inputs = tokenizer.encode("The Los Angeles Lakers", return_tensors='pt').to(model.device)
# set top_k = 50 and set top_p = 0.95 and num_return_sequences = 3
outputs = model.generate(
    inputs,
    do_sample=True, 
    max_length=50, 
    top_k=50, 
    top_p=0.95, 
    temperature=1,
    num_return_sequences=3
)

print("Output:\n" + 100 * '-')
for i, sample_output in enumerate(outputs):
    print("{}: {}".format(i, tokenizer.decode(sample_output, skip_special_tokens=True)).replace('\n',' '))

Output:
----------------------------------------------------------------------------------------------------
0: The Los Angeles Lakers have won four in a row, including a 5-1 win on Sunday against the Los Angeles Clippers.  "I think everyone wants to believe in him. I think we can come back at the same time after this game
1: The Los Angeles Lakers won it all by beating the Chicago Bulls 3-2 and advancing to the Finals, a game that they beat again on December 14, 1994.  They were defeated in overtime by the Los Angeles Lakers at Oracle Arena, but
2: The Los Angeles Lakers have lost just once in their last seven games and have only been outscored by 17.6 points, a discrepancy that ranks fifth in the NBA. They've also gone 18-5 after losing eight in a row on the


## 2.3 Fine-Tuning GPT-2
Okay now time for the best part!

Generating general-purpose text from pre-trained models is great, but what if we want our text to be in a specific genre or style? Luckily for us, the GPT family of models use the idea of "Transfer learning" -- using knowledge gained from one problem (or training setting), and applying it to another area or domain. The idea of transfer learning for NLP, is that we can train a language model on general texts, and then adapt it to use it for a specific task or domain that we're interested in. This process is also called **fine-tuning**.

In this section we'll walk you through an example of using HuggingFace to fine-tune GPT-2 and then you'll be asked to fine-tune GPT-2 on two datasets of your own choosing!

### Fine-Tuning Example using HuggingFace Datasets library: Crime and Punishment

For our fine-tuning example we're going to train GPT-2 to mimic the style of Fyodor Dostoevsky's novel "Crime and Punishment"

We will be downloading our data using the HuggingFace [Datasets](https://huggingface.co/docs/datasets/) library.

In [19]:
!pip install datasets

In [20]:
from transformers import Trainer, TrainingArguments, DataCollatorForLanguageModeling
import datasets
from datasets import load_dataset, list_datasets

### Step 1: Initialize a Brand New GPT-2 Model and Tokenizer

In [21]:
tokenizer = AutoTokenizer.from_pretrained('gpt2')
tokenizer.pad_token = tokenizer.eos_token
model = AutoModelForCausalLM.from_pretrained('gpt2').cuda()
data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)

In [22]:
data_collator

DataCollatorForLanguageModeling(tokenizer=PreTrainedTokenizerFast(name_or_path='gpt2', vocab_size=50257, model_max_len=1024, is_fast=True, padding_side='right', special_tokens={'bos_token': '<|endoftext|>', 'eos_token': '<|endoftext|>', 'unk_token': '<|endoftext|>', 'pad_token': '<|endoftext|>'}), mlm=False, mlm_probability=0.15)

###Step 2: Load the text of "Crime and Punishment" and tokenize it

The 'load_dataset' function queries for a dataset with a certain tag and downloads the corresponding data from HuggingFace's hosting site. This allows us to download all sorts of datasets through the same interface!

The documentation for load_dataset can be found [here](https://huggingface.co/docs/datasets/package_reference/loading_methods.html#datasets.load_dataset)

Here we take our tokenizer and run it on the entirety of Crime and Punishment in a single batch by using map on our custom encode function.

In [23]:
def encode(batch): 
    return tokenizer([x.strip('\n\r') for x in batch['line']], truncation=True, padding=True)

In [24]:
crime_and_punishment = load_dataset('crime_and_punish', split='train')

Reusing dataset crime_and_punish (/root/.cache/huggingface/datasets/crime_and_punish/crime-and-punish/1.0.0/87ec36ba9cb8741325bea3e40a6d4525210c8f8ef13e7b07872fe32eb72c13ac)


In [25]:
crime_and_punishment

Dataset({
    features: ['line'],
    num_rows: 21969
})

In [26]:
processed_dataset = crime_and_punishment.map(encode, batched=True, batch_size=len(crime_and_punishment))

Loading cached processed dataset at /root/.cache/huggingface/datasets/crime_and_punish/crime-and-punish/1.0.0/87ec36ba9cb8741325bea3e40a6d4525210c8f8ef13e7b07872fe32eb72c13ac/cache-98f25f7c4eac82b8.arrow


In [27]:
processed_dataset

Dataset({
    features: ['attention_mask', 'input_ids', 'line'],
    num_rows: 21969
})

In [28]:
processed_dataset.set_format('torch', columns=['input_ids', 'attention_mask'])

### Step 3: Initialize the Trainer

The 'Trainer' module is the main way we perform fine-tuning. In order to initialize a Trainer, you need a model, tokenizer, TrainingArguments, your training data (in a Dataset object) and something called a data_collator (which tells the Trainer not to look for a vector of labels). 

In [29]:
training_args = TrainingArguments(
    output_dir='/content/',
    overwrite_output_dir=True,
    num_train_epochs=1,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=64,
    logging_steps=100,
    weight_decay=0.01,
    logging_dir='./logs',
)

In [30]:
training_args

TrainingArguments(output_dir=/content/, overwrite_output_dir=True, do_train=False, do_eval=None, do_predict=False, evaluation_strategy=IntervalStrategy.NO, prediction_loss_only=False, per_device_train_batch_size=16, per_device_eval_batch_size=64, gradient_accumulation_steps=1, eval_accumulation_steps=None, learning_rate=5e-05, weight_decay=0.01, adam_beta1=0.9, adam_beta2=0.999, adam_epsilon=1e-08, max_grad_norm=1.0, num_train_epochs=1, max_steps=-1, lr_scheduler_type=SchedulerType.LINEAR, warmup_ratio=0.0, warmup_steps=0, logging_dir=./logs, logging_strategy=IntervalStrategy.STEPS, logging_first_step=False, logging_steps=100, save_strategy=IntervalStrategy.STEPS, save_steps=500, save_total_limit=None, no_cuda=False, seed=42, fp16=False, fp16_opt_level=O1, fp16_backend=auto, fp16_full_eval=False, local_rank=-1, tpu_num_cores=None, tpu_metrics_debug=False, debug=False, dataloader_drop_last=False, eval_steps=100, dataloader_num_workers=0, past_index=-1, run_name=/content/, disable_tqdm=F

In [31]:
trainer = Trainer(
    model=model,
    tokenizer=tokenizer,
    args=training_args,
    data_collator=data_collator,
    train_dataset=processed_dataset,
)

### Step 4: Fine-Tune the Model!

Now we're done! All we have to do is hit run and sit back!

In [32]:
trainer.train()

TrainOutput(global_step=1374, training_loss=3.5791839199815776, metrics={'train_runtime': 223.7963, 'train_samples_per_second': 6.14, 'total_flos': 574101809809920.0, 'epoch': 1.0, 'init_mem_cpu_alloc_delta': 0, 'init_mem_gpu_alloc_delta': 0, 'init_mem_cpu_peaked_delta': 0, 'init_mem_gpu_peaked_delta': 0, 'train_mem_cpu_alloc_delta': 10391552, 'train_mem_gpu_alloc_delta': 1499693056, 'train_mem_cpu_peaked_delta': 11399168, 'train_mem_gpu_peaked_delta': 1278872064})

### Step 5: Save the Model and use it to Generate!

Save your fine-tuned model and compare its output with regular GPT-2's output to see the difference for yourself!

In [33]:
trainer.save_model('./dostoevskypt2')

In [34]:
dostoevskypt2 = pipeline('text-generation', model='./dostoevskypt2', device=0)
gpt2 = pipeline('text-generation', model='gpt2', device=0)

In [35]:
print(dostoevskypt2('Saint Petersburg is'))
print(gpt2('Saint Petersburg is'))

[{'generated_text': 'Saint Petersburg is haunted by the same strange feeling of terror, of the same peculiar feeling of being alone. “When will you come in?” he asked abruptly at last. “I am sorry,” he looked more carefully,'}]
[{'generated_text': 'Saint Petersburg is an extremely important centre – the oldest city in Russia – one of the top five in the world and one of the second-best centers of science, technology, engineering, and mathematics in the world. It boasts an amazing population of nearly'}]


## PERPLEXITY

(3 pts) Using the pointer [here](https://huggingface.co/transformers/perplexity.html), compute the perplexity of the GPT2 pre-trained model on the Wikipedia test set (you can keep the same hyperparameters as in the link) 

In [36]:
import torch
from tqdm import tqdm

In [37]:
test_wiki = load_dataset('wikitext', 'wikitext-2-raw-v1', split='test')

Reusing dataset wikitext (/root/.cache/huggingface/datasets/wikitext/wikitext-2-raw-v1/1.0.0/47c57a6745aa5ce8e16a5355aaa4039e3aa90d1adad87cef1ad4e0f29e74ac91)


In [38]:
def compute_perplexity(model, tokenizer, test_data, format='text'):
    encodings = tokenizer('\n\n'.join(test_data[format]), return_tensors='pt')
    max_length = model.config.n_positions
    stride = 512

    lls = []
    for i in tqdm(range(0, encodings.input_ids.size(1), stride)):
        begin_loc = max(i + stride - max_length, 0)
        end_loc = min(i + stride, encodings.input_ids.size(1))
        trg_len = end_loc - i    # may be different from stride on last loop
        input_ids = encodings.input_ids[:,begin_loc:end_loc].to(model.device)
        target_ids = input_ids.clone()
        target_ids[:,:-trg_len] = -100

        with torch.no_grad():
            outputs = model(input_ids, labels=target_ids)
            log_likelihood = outputs[0] * trg_len

        lls.append(log_likelihood)

    ppl = torch.exp(torch.stack(lls).sum() / end_loc)
    logger.info("perplex:{}".format(ppl))

In [39]:
## YOUR CODE HERE - FOR COMPUTING PERPLEXITY OF GPT2 ON WIKIPEDIA TEST SET 
compute_perplexity(gpt2.model, gpt2.tokenizer, test_wiki)

Token indices sequence length is longer than the specified maximum sequence length for this model (287644 > 1024). Running this sequence through the model will result in indexing errors
100%|██████████| 562/562 [00:19<00:00, 28.50it/s]
perplex:25.170494079589844


> Perplexity = 25.17

(2 pts) Compute the  perplexity of the dostoevskypt2 model on Wikipedia test set

In [40]:
## YOUR CODE HERE - FOR COMPUTING PERPLEXITY OF DOSTOEVSKYPT2 ON WIKIPEDIA TEST SET
compute_perplexity(dostoevskypt2.model, dostoevskypt2.tokenizer, test_wiki)

Token indices sequence length is longer than the specified maximum sequence length for this model (287644 > 1024). Running this sequence through the model will result in indexing errors
100%|██████████| 562/562 [00:19<00:00, 28.48it/s]
perplex:65.72527313232422


> Perplexity = 65.72

(2 pts) Compute the perplexity of the GPT2 pre-trained model on the Crime and Punishment train dataset

In [41]:
## YOUR CODE HERE - FOR COMPUTING PERPLEXITY OF GPT2 ON CRIME AND PUNISHMENT TRAIN DATASET 
compute_perplexity(gpt2.model, gpt2.tokenizer, crime_and_punishment, 'line')

100%|██████████| 705/705 [00:24<00:00, 28.42it/s]
perplex:83.48881530761719


> Perplexity = 83.48

(2 pts) Compute the **train** perplexity of the **dostoevskypt2** model 




In [42]:
## YOUR CODE HERE - FOR COMPUTING PERPLEXITY OF DOSTOEVSKYPT2 ON CRIME AND PUNISHMENT TRAIN DATASET 
compute_perplexity(dostoevskypt2.model, dostoevskypt2.tokenizer, crime_and_punishment, 'line')

100%|██████████| 705/705 [00:24<00:00, 28.35it/s]
perplex:58.41886901855469


> Perplexity = 58.41

(2 pts) Compute the perplexity of the GPT2 model on your Pride and Prejudice text

In [43]:
## YOUR CODE HERE - FOR COMPUTING PERPLEXITY OF GPT2 ON PRIDE AND PREJUDICE TEXT 
lines = []
with open('prideAndPrejudice.txt') as f:
    for line in f:
        lines.append(line.strip())
pride_and_prejudice={'line' : lines}

In [44]:
compute_perplexity(gpt2.model, gpt2.tokenizer, pride_and_prejudice, 'line')

100%|██████████| 304/304 [00:10<00:00, 28.44it/s]
perplex:27.930387496948242


> Perplexity = 27.93

(2 pts) Compute the perplexity of the dostoevskypt2 model on your Pride and Prejudice text

In [45]:
## YOUR CODE HERE - FOR COMPUTING PERPLEXITY OF DOSTOEVSKYPT2 ON PRIDE AND PREJUDICE TEXT 
compute_perplexity(dostoevskypt2.model, dostoevskypt2.tokenizer, pride_and_prejudice, 'line')

100%|██████████| 304/304 [00:10<00:00, 28.49it/s]
perplex:40.802947998046875


> Perplexity = 40.80

1. (1 pt) Which model performs better on Wikipedia text?
   > GPT2
2. (1 pt) Which model performs better on Pride and Prejudice text?
   > GPT2

### Now's Your Turn!

**Your job is to fine-tune GPT2 one more time with your choice of fine-tuning dataset.**

*****For the fine-tuned model you create, you should clearly demonstrate (through visible generation outputs and analysis) that your fine-tuned model follows the desired style better than vanilla GPT2** ***

Please make sure to give a brief description 

In order to see which datasets are available for download, run the cell below. Pick one that you think would be interesting!

In [46]:
datasets_list = list_datasets()
print(', '.join(dataset for dataset in datasets_list[:10]))

acronym_identification, ade_corpus_v2, adversarial_qa, aeslc, afrikaans_ner_corpus, ag_news, ai2_arc, air_dialogue, ajgt_twitter_ar, allegro_reviews


### Tips
- Most of the datasets hosted by HuggingFace are not meant for Causal LM fine-tuning. Make sure you preprocess them accordingly if you want to use them.
- In order to check out information about a dataset hosted by huggingface you can use [this web viewer](https://huggingface.co/datasets/viewer/?dataset=crime_and_punish). Try to avoid downloading a dataset that's too big!
- You will likely have to change the custom 'encode' function for each new dataset you want to fine-tune on. You need to change batch['line'] to instead index with the correct column label for your specific dataset (it probably wont be called 'line').

### Useful Links
[load_datasets Documentation](https://huggingface.co/docs/datasets/package_reference/loading_methods.html#datasets.load_dataset)

[Trainer Documentation](https://huggingface.co/transformers/main_classes/trainer.html#id1)

[Example: Fine-Tuning BERT for Esperanto](https://colab.research.google.com/github/huggingface/blog/blob/master/notebooks/01_how_to_train.ipynb#scrollTo=zTgWPa9Dipk2)

[Example: Fine-Tuning for IMDb Classification](https://colab.research.google.com/drive/1-JIJlao4dI-Ilww_NnTc0rxtp-ymgDgM?usp=sharing#scrollTo=5DEWNilys9Ty)


#### Dataset \#1

> Use ag_news dataset

In [47]:
## YOUR CODE HERE - FOR FINE-TUNING GPT2 ON DATASET 
from transformers import Trainer, TrainingArguments, DataCollatorForLanguageModeling
import datasets
from datasets import load_dataset, list_datasets
from transformers import AutoTokenizer, AutoModelForCausalLM

In [48]:
tokenizer = AutoTokenizer.from_pretrained('gpt2')
tokenizer.pad_token = tokenizer.eos_token
model = AutoModelForCausalLM.from_pretrained('gpt2').cuda()
data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)

In [49]:
ag_news = load_dataset('ag_news', split='test')

Using custom data configuration default
Reusing dataset ag_news (/root/.cache/huggingface/datasets/ag_news/default/0.0.0/fb5c5e74a110037311ef5e904583ce9f8b9fbc1354290f97b4929f01b3f48b1a)


In [50]:
ag_news

Dataset({
    features: ['text', 'label'],
    num_rows: 7600
})

In [51]:
def encode(batch): 
    return tokenizer([x.strip('\n\r') for x in batch['text']], truncation=True, padding=True)

In [52]:
processed_dataset = ag_news.map(encode, batched=True, batch_size=len(ag_news))

Loading cached processed dataset at /root/.cache/huggingface/datasets/ag_news/default/0.0.0/fb5c5e74a110037311ef5e904583ce9f8b9fbc1354290f97b4929f01b3f48b1a/cache-ac159b94f154fdc2.arrow


In [53]:
processed_dataset.set_format('torch', columns=['input_ids', 'attention_mask'])

In [54]:
processed_dataset.remove_columns_('label')

remove_columns_ is deprecated and will be removed in the next major version of datasets. Use the dataset.remove_columns method instead.


In [55]:
processed_dataset

Dataset({
    features: ['attention_mask', 'input_ids', 'text'],
    num_rows: 7600
})

In [56]:
training_args = TrainingArguments(
    output_dir='/content/',
    overwrite_output_dir=True,
    num_train_epochs=1,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=32,
    logging_steps=100,
    weight_decay=0.01,
    logging_dir='./log2',
)

In [57]:
trainer = Trainer(
    model=model,
    tokenizer=tokenizer,
    args=training_args,
    data_collator=data_collator,
    train_dataset=processed_dataset,
)

In [58]:
trainer.train()

TrainOutput(global_step=950, training_loss=3.6390905279862253, metrics={'train_runtime': 252.9898, 'train_samples_per_second': 3.755, 'total_flos': 1469683908403200.0, 'epoch': 1.0, 'init_mem_cpu_alloc_delta': 0, 'init_mem_gpu_alloc_delta': 0, 'init_mem_cpu_peaked_delta': 0, 'init_mem_gpu_peaked_delta': 0, 'train_mem_cpu_alloc_delta': -146526208, 'train_mem_gpu_alloc_delta': 1501162496, 'train_mem_cpu_peaked_delta': 167673856, 'train_mem_gpu_peaked_delta': 4862874112})

(3 pts) YOUR ANSWER HERE - BRIEF DESCRIPTION OF THE DATASET YOU CHOSE
> AG is a collection of more than 1 million news articles. News articles have been gathered from more than 2000 news sources by ComeToMyHead in more than 1 year of activity.

In [59]:
## YOUR CODE HERE - FOR GENERATION WITH YOUR FINE-TUNED MODEL AND COMPARISON WITH REGULAR GPT2
from transformers import pipeline
trainer.save_model('./ag_news_model')
ag_news_model = pipeline('text-generation', model='./ag_news_model', device=0)
gpt2 = pipeline('text-generation', model='gpt2', device=0)

In [60]:
print(ag_news_model('The cat is so cute'))
print(gpt2('The cat is so cute'))

[{'generated_text': 'The cat is so cute, don #39;t you think? What if a great cat is so perfect that the humans don #39;t even care about them? After it was just one of those years of living in a place that could'}]
[{'generated_text': 'The cat is so cute and cute and adorable that it\'s so cute, even when it\'s just for decoration, it\'s so cute and adorable," said David.\n\nOne of the more memorable parts of the ride is seeing the cat come to'}]


(4 pts) YOUR ANSWER HERE - COMPARISON OF YOUR DATASET'S FINE-TUNED OUTPUT VS NON-FINE-TUNED OUTPUT 
> Due to fine-tuned on news dataset, the generated text is more like news, while the non-fine-tuned output is more general. We could generate specific sentences by using fine-tuned model.
